admin/billing/orb.go

```diff
@@ -0,0 +1,377 @@
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *Orb) ReportUsage(ctx context.Context, usage []*Usage) error {
+	var orbUsage []orb.EventIngestParamsEvent
+	// sync max 500 events at a time
+	for _, u := range usage {
+		eventName := u.MetricName + "_" + string(u.ReportingGrain)
+		// use end time minus 1 second to make sure the event is attributed to the current time bucket
+		eventTime := u.EndTime.Add(-1 * time.Second)
+		// generate idempotency key using customer id, timestamp, event name and metadata
+		idempotencyKey := fmt.Sprintf("%s_%d_%s_%v", u.CustomerID, eventTime.UnixMilli(), eventName, u.Metadata)
+
+		props := make(map[string]interface{}, len(u.Metadata)+1)
+		for k, v := range u.Metadata {
+			props[k] = v
+		}
+		props["amount"] = u.Amount
+
+		orbUsage = append(orbUsage, orb.EventIngestParamsEvent{
+			ExternalCustomerID: orb.String(u.CustomerID),
+			EventName:          orb.String(eventName),
+			IdempotencyKey:     orb.String(idempotencyKey),
+			Timestamp:          orb.F(eventTime),
+			Properties:         orb.F[any](props),
+		})
+
+		if len(orbUsage) == requestMaxLimit {
+			err := o.pushUsage(ctx, &orbUsage)
+			if err != nil {
+				return err
+			}
+			orbUsage = nil
+		}
+	}
+
+	if len(orbUsage) > 0 {
+		err := o.pushUsage(ctx, &orbUsage)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *Orb) pushUsage(ctx context.Context, usage *[]orb.EventIngestParamsEvent) error {
+	re := retrier.New(retrier.ExponentialBackoff(5, 500*time.Millisecond), retryErrClassifier{})
+	err := re.RunCtx(ctx, func(ctx context.Context) error {
+		resp, err := o.client.Events.Ingest(ctx, orb.EventIngestParams{
+			Events: orb.F(*usage),
+		})
+		if err != nil {
+			return err
+		}
+		if len(resp.ValidationFailed) > 0 {
+			errMsg := fmt.Sprintf("validation failure for %d events, showing first few:", len(resp.ValidationFailed))
+			for i := 0; i < 5 && i < len(resp.ValidationFailed); i++ {
+				errMsg += fmt.Sprintf("\n%s: %s", resp.ValidationFailed[i].IdempotencyKey, resp.ValidationFailed[i].ValidationErrors)
+			}
+			return errors.New(errMsg)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *Orb) GetReportingGranularity() UsageReportingGranularity {
+	return UsageReportingGranularityHour
+}
+
+func (o *Orb) GetReportingWorkerCron() string {
+	// run every hour at around end of the hour
+	return "55 * * * *"
+}
+
+func (o *Orb) getAllPlans(ctx context.Context) ([]*Plan, error) {
+	plans, err := o.client.Plans.List(ctx, orb.PlanListParams{
+		Limit:  orb.Int(requestMaxLimit), // TODO handle pagination, for now don't expect more than 500 plans
+		Status: orb.F(orb.PlanListParamsStatusActive),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var billingPlans []*Plan
+	for i := 0; i < len(plans.Data); i++ {
+		billingPlan, err := getBillingPlanFromOrbPlan(&plans.Data[i])
+		if err != nil {
+			return nil, err
+		}
+		billingPlans = append(billingPlans, billingPlan)
+	}
```

Comment:
Should we maybe filter out those plans where `Name == ""`? In case we forget to set it in Orb, then could someone send a manual API call to upgrade to `""` and then get access to some internal plan? Or is there something that protects against that?

Reply:
Good catch but it cannot happen currently, in the API where `GetPlanByName` is used, a check exists to disallow empty string. Also there is option to add plan in the cli cmd using biller plan id as well (not sure if we need that), so filtering would cause that to fail. 

But to be more defensive I can return err from `GetPlanByName` if name is empty ?

Reply:
Sounds good with the additional defensive option